test/hummingbot/connector/exchange/kucoin/test_kucoin_exchange.py

```diff
@@ -418,3 +294,1329 @@ def test_client_order_id_on_order(self, mocked_nonce):
+        self.exchange._set_current_timestamp(1640780000)
+
+        self.exchange.start_tracking_order(
+            order_id="OID1",
+            exchange_order_id="4",
+            trading_pair=self.trading_pair,
+            trade_type=TradeType.BUY,
+            price=Decimal("10000"),
+            amount=Decimal("100"),
+            order_type=OrderType.LIMIT,
+        )
+
+        self.assertIn("OID1", self.exchange.in_flight_orders)
+        order = self.exchange.in_flight_orders["OID1"]
+
+        url = web_utils.rest_url(f"{CONSTANTS.ORDERS_PATH_URL}/{order.exchange_order_id}")
+        regex_url = re.compile(f"^{url}".replace(".", r"\.").replace("?", r"\?"))
+
+        mock_api.delete(regex_url,
+                        status=400,
+                        callback=lambda *args, **kwargs: request_sent_event.set())
+
+        self.exchange.cancel(trading_pair=self.trading_pair, order_id="OID1")
+        self.async_run_with_timeout(request_sent_event.wait())
+
+        cancel_request = next(((key, value) for key, value in mock_api.requests.items()
+                               if key[1].human_repr().startswith(url)))
+        self._validate_auth_credentials_present(cancel_request[1][0])
+
+        self.assertEquals(0, len(self.order_cancelled_logger.event_log))
+
+        self.assertTrue(
+            self._is_logged(
+                "NETWORK",
+                f"Failed to cancel order {order.client_order_id}: Error executing request DELETE "
+                f"/api/v1/orders/{order.exchange_order_id}. HTTP status is 400. Error: "
+            )
+        )
+
+    def test_cancel_order_without_exchange_order_id_marks_order_as_fail_after_retries(self):
+        update_event = MagicMock()
+        update_event.wait.side_effect = asyncio.TimeoutError
+
+        self.exchange._set_current_timestamp(1640780000)
+
+        self.exchange.start_tracking_order(
+            order_id="OID1",
+            exchange_order_id=None,
+            trading_pair=self.trading_pair,
+            trade_type=TradeType.BUY,
+            price=Decimal("10000"),
+            amount=Decimal("100"),
+            order_type=OrderType.LIMIT,
+        )
+
+        self.assertIn("OID1", self.exchange.in_flight_orders)
+        order = self.exchange.in_flight_orders["OID1"]
+        order.exchange_order_id_update_event = update_event
+
+        self.async_run_with_timeout(self.exchange._execute_cancel(
+            trading_pair=order.trading_pair,
+            order_id=order.client_order_id,
+        ))
+
+        self.assertEqual(0, len(self.order_cancelled_logger.event_log))
+
+        self.assertTrue(
+            self._is_logged(
+                "WARNING",
+                f"Failed to cancel the order {order.client_order_id} because it does not have an exchange order id yet"
+            )
+        )
+
+        # After the fourth time not finding the exchange order id the order should be marked as failed
+        for i in range(4):
```

Comment:
I think this should be set in `CONSTANTS`

Reply:
Done